File: 202. Happy Number/solution_3.java
//非happy数会有一个特点，在循环过程中会出现4这个值
class Solution {
    public boolean isHappy(int n) {
        while (true) {
            if (n == 4) return false;
            if (n == 1) return true;
            n = getSum(n);
        }
    }
    private int getSum(int n) {
        int sum = 0;
        while (n != 0) {
            sum += Math.pow(n % 10, 2);
            n /= 10;
        }
        return sum;
    }
}

File: 96. Unique Binary Search Trees/solution_3.cpp
class Solution {
public:
    int numTrees(int n) {
        vector<int> vec(n + 1);
        vec[0] = 1;
        vec[1] = 1;
        for (int i = 2; i <= n; ++ i) {
            for (int j = 1; j <= i; ++ j)
                vec[i] += vec[j - 1] * vec[i - j];
        }
        return vec[n];
    }
}; 
//g(i) 表示序列长度为i的树个数
//f(j, i) 表示序列长度为i， 以第j个节点作为根的树的个数
//g(i) = sum_(j = 1)^(i)f(j, i)
//f(j, i) = g(j - 1)*g(i - j)
//所以 g(i) = sum_(j = 1)^(i)g(j - 1)*g(i - j)